compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -81,6 +81,28 @@ object Inferencing {
     if (depVars.nonEmpty) instantiateSelected(tp, depVars.toList)
   }
 
+  /** If `tp` is top-level type variable with a lower bound in the current constraint,
+   *  instantiate it from below. We also look for TypeVars whereever their instantiation
+   *  could uncover new type members.
+   */
+  def couldInstantiateTypeVar(tp: Type)(using Context): Boolean = tp.dealias match
+    case tvar: TypeVar
+    if !tvar.isInstantiated
+       && ctx.typerState.constraint.contains(tvar)
+       && tvar.hasLowerBound =>
+      tvar.instantiate(fromBelow = true)
```

Comment:
When trying to select from something that doesn't have a lower-bound, rather than giving up I think it makes sense to try the upper bound, since it might be precise enough, also we could take `instDireciton` into account to be closer to the normal behavior of `isFullyDefined`:
```suggestion
      =>
      val direction = instDirection(tvar.origin)
      tvar.instantiate(fromBelow = direction < 0 || tvar.hasLowerBound)
```

Reply:
I am not sure that will help. The situation is that we try to select a member that does not exist on the type variable. Instantiating to the upper bound would not help in this case. If the member existed on the upper bound, it would also exist on the variable itself.

Reply:
> If the member existed on the upper bound, it would also exist on the variable itself.

Ah indeed, because `goParam` in `findMember` will recurse on the upper bound, but isn't that actually problematic? If I have a selection `x.foo` where the type of `x` is a type variable `?T <: B`, we'll end up selecting some overload of `B#foo`, but it's possible that `?T` is then instantiated to some more precise type which contains another overload of `foo` (or an override with a more precise result type), meaning that the member selected is not preserved by retyping (in practice our ReTyper will keep the original selection, but this still seems like a source of confusion since the meaning of the code is not apparent from the types the user sees).

Reply:
This could be a problem. But (1) it's not related to this issue.(2) I don't know how to solve it.

Reply:
I think the change I'm proposing together with removing the recursion on the upper bound in `goParam` would solve that issue, but I agree this can be treated separately from this PR.